src/connection-manager/connections.ts

```diff
@@ -107,41 +123,58 @@ async function setHandlers(
       }
     })
     ws.on('close', async (code, reason) => {
-      const nodeNetworks = networks ?? 'unknown network'
-      if (connectionsInitialized) {
-        log.error(
-          `Websocket closed for ${
-            ws.url
-          } on ${nodeNetworks} with code ${code} and reason ${reason.toString(
-            'utf-8',
-          )}.`,
+      log.error(
+        `Websocket closed for ${ws.url} on ${
+          networks ?? 'unknown network'
+        } with code ${code} and reason ${reason.toString('utf-8')}.`,
+      )
+
+      const delay = BASE_RETRY_DELAY * 2 ** retryCount
+
+      if (CLOSING_CODES.includes(code) && delay <= MAX_RETRY_DELAY) {
+        log.info(
+          `Reconnecting to ${ws.url} on ${
+            networks ?? 'unknown network'
+          } after ${delay}ms...`,
         )
-        if (CLOSING_CODES.includes(code)) {
-          log.info(
-            `Reconnecting to ${ws.url} on ${networks ?? 'unknown network'}...`,
-          )
+        // Clean up the old Websocket connection
+        connections.delete(ws.url)
+        ws.terminate()
+        resolve()
+
```

Comment:
```suggestion
void saveNodeWsUrl(ws.url, false)
```

Since the websocket is closed, that should be recorded in the database. If it reopens successfully, then there will be another update to the database inside `ws.on('open', ...)`

Reply:
This is missing in the original code base too

Reply:
This line causes deadlock since we swamp the database with many closed connection. The suggestion to add new column/table to track the health of the websocket connection is a good idea. We can store those health periodically.